Lib/subprocess.py

```diff
@@ -1096,13 +1096,17 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
 
             assert not pass_fds, "pass_fds not supported on Windows."
 
-            if not isinstance(args, str):
-                try:
-                    args = os.fsdecode(args)  # os.PathLike -> str
-                except TypeError:  # not an os.PathLike, must be a sequence.
-                    args = list(args)
-                    args[0] = os.fsdecode(args[0])  # os.PathLike -> str
-                    args = list2cmdline(args)
+            if isinstance(args, str):
+                pass
+            elif not shell and isinstance(args, os.PathLike):
+                args = list2cmdline([os.fsdecode(args)])
+            else:
+                args = list(args)
+                args[0] = os.fsdecode(args[0])
+                args = list2cmdline(args)
```

Comment:
This needs to be addressed in `list2cmdline` to be consistent with how POSIX converts every element of `args` via `PyUnicode_FSConverter`.  For example:

    if isinstance(args, bytes):
        args = os.fsdecode(args)
    elif isinstance(args, os.PathLike):
        if shell:
            raise ValueError('PathLike args is not allowed when '
                             'shell is true.')
        else:
            args = [args]

    if not isinstance(args, str):
        args = list2cmdline(args)

    if executable is not None:
        executable = os.fsdecode(executable)

Then in `list2cmdline`, decode the items via `os.fsdecode` :

    for arg in seq:
        if not isinstance(arg, str):
            arg = os.fsdecode(arg)

Reply:
This does look like it would be nicer, allowing more mixing and matching of types.

Reply:
@eryksun 's comment is related to the old version of the PR. I addressed it, except that that I decided to reject bytes `args` on Windows with shell=True as it is ambiguous and is not actually needed. It is enough to have a difference in interpreting str `args`.